src/mudlet.cpp

```diff
@@ -2335,37 +2331,53 @@ void mudlet::slot_replaySpeedDown()
     replaySpeedDisplay->show();
 }
 
+/* loop through and stop all sounds */
 void mudlet::stopSounds()
 {
-    mpMusicBox1->stop();
-    mpMusicBox2->stop();
-    mpMusicBox3->stop();
-    mpMusicBox4->stop();
+    QListIterator<QMediaPlayer *> itMusicBox( MusicBoxList );
+    
+    while( itMusicBox.hasNext() ) {
+        itMusicBox.next()->stop();
+    }
 }
 
-void mudlet::playSound( QString s )
+void mudlet::playSound( QString s, int soundVolume )
 {
-    if( mpMusicBox1->state() != QMediaPlayer::PlayingState )
-    {
-        mpMusicBox1->setMedia( QUrl::fromLocalFile( s ) );
-        mpMusicBox1->play();
-    }
-    else if( mpMusicBox2->state() != QMediaPlayer::PlayingState )
-    {
-        mpMusicBox2->setMedia( QUrl::fromLocalFile( s ) );
-        mpMusicBox2->play();
-    }
-    else if( mpMusicBox3->state() != QMediaPlayer::PlayingState )
-    {
-        mpMusicBox3->setMedia( QUrl::fromLocalFile( s ) );
-        mpMusicBox3->play();
+    QListIterator<QMediaPlayer *> itMusicBox( MusicBoxList );
+    QMediaPlayer * pPlayer=0;
+    
+    /* find first available inactive QMediaPlayer */
+    while( itMusicBox.hasNext() ) {
+        QMediaPlayer * pTestPlayer=itMusicBox.next();
+        
+        if ( pTestPlayer->state() != QMediaPlayer::PlayingState && pTestPlayer->mediaStatus() != QMediaPlayer::LoadingMedia ) {
+            pPlayer=pTestPlayer;
+            break;
+        }
     }
-    else
-    {
-        mpMusicBox4->stop();
-        mpMusicBox4->setMedia( QUrl::fromLocalFile( s ) );
-        mpMusicBox4->play();
+
+    /* no available QMediaPlayer, create a new one */
+    if (!pPlayer) {
+        pPlayer = new QMediaPlayer(this);
+
+        if (!pPlayer) {
+            /* It (should) be impossible to ever reach this */
+            Host * pH = getActiveHost();
+            
+            if (pH) {
+                pH->postMessage( "\n[  ERROR  ]  - Unable to create new QMediaPlayer object\n" );    
+            }
+            
+            return;
+        }
+
+        MusicBoxList.append(pPlayer);
     }
+
+    /* set volume and play sound */
+    pPlayer->setMedia( QUrl::fromLocalFile( s ) );
+    pPlayer->setVolume(soundVolume);
+    pPlayer->play();
```

Comment:
Might I suggest this replacement for the above that better fits our current "standards" for Lua commands with better error reporting?

```
int TLuaInterpreter::playSoundFile( lua_State * L )
{
    Host * pHost = TLuaInterpreter::luaInterpreterMap[L];
    if( ! pHost ) {
        lua_pushnil( L );
        lua_pushstring( L, tr( "playSoundFile: NULL Host pointer - something is wrong!" )
                        .toUtf8().constData() );
        return 2;
    }

    QString sound;
    if( ! lua_isstring( L, 1 ) ) {
        lua_pushstring( L, tr( "playSoundFile: bad argument #1 type (sound pathAndFilename as string expected,\n"
                               "got %1!)" )
                        .arg( luaL_typename( L, 1) )
                        .toUtf8().constData() );
        lua_error( L );
        return 1;
    }
    else {
        // Windows platforms, at least, can have non-ASCII characters in path/filenames:
        sound = QString::fromUtf8( lua_tostring( L, 1 );
 
        if( sound.isEmpty() {
            lua_pushnil( L );
            lua_pushstring( L, tr( "playSoundFile: bad argument #1 value (sound pathAndFilename cannot be empty.)" )
                            .toUtf8().constData() );
            return 2;
        }

#ifdef Q_OS_WIN32
        sound.replace( QLatin1Char( '\\' ), QLatin1Char( '/' ) );
#endif
    }

    qint64 soundVolume = 100;
    if( lua_gettop( L ) > 1 ) {
        if( ! lua_isnumber( L, 2 ) {
            lua_pushstring( L, tr( "playSoundFile: bad argument #2 type (sound volume percentage as number is\n"
                                   "optional {default = 100 percent if ommitted}, got %1!)"
                            .arg( luaL_typename( L, 2 ) )
                            .toUtf8().constData() );
            lua_error( L );
            return 1;
        }

        soundVolume = lua_tointeger( L, 2 );
        if( soundVolume < 0 || soundVolume > 100 ) {
            lua_pushnil( L );
            lua_pushstring( L, tr( "setAreaName: bad argument #2 value (sound volume %1 is not in range"
                                   "{0 = silent; 100 = full, default if omitted}.)" )
                            .arg( soundVolume )
                            .toUtf8().constData() );
            return 2;
        }
    }

    mudlet::self()->playSound( sound, soundVolume );
    lua_pushboolean( L, 1 );
    return 1;
}
```